vsphere/datadog_checks/vsphere/mor_cache.py

```diff
@@ -0,0 +1,121 @@
+# (C) Datadog, Inc. 2010-2017
+# All rights reserved
+# Licensed under Simplified BSD License (see LICENSE)
+import threading
+import time
+
+
+class MorNotFoundError(Exception):
+    pass
+
+
+class MorCache:
+    """
+    Implements a thread safe storage for Mor objects.
+    For each instance key, the cache maps: mor_name --> mor_dict_object
+    """
+    def __init__(self):
+        self._mor = {}
+        self._mor_lock = threading.RLock()
+
+    def init_instance(self, key):
+        """
+        Create an empty instance if it doesn't exist.
+        If the instance already exists, this is a noop.
+        """
+        with self._mor_lock:
+            if key not in self._mor:
+                self._mor[key] = {}
+
+    def contains(self, key):
+        """
+        Return whether an instance key is present.
+        """
+        with self._mor_lock:
+            return key in self._mor
+
+    def instance_size(self, key):
+        """
+        Return how many Mor objects are stored for the given instance.
+        If the key is not in the cache, raises a KeyError.
+        """
+        with self._mor_lock:
+            return len(self._mor[key])
+
+    def set_mor(self, key, name, mor):
+        """
+        Store a Mor object in the cache with the given name.
+        If the key is not in the cache, raises a KeyError.
+        """
+        with self._mor_lock:
+            self._mor[key][name] = mor
+            self._mor[key][name]['creation_time'] = time.time()
+
+    def get_mor(self, key, name):
+        """
+        Return the Mor object identified by `name` for the given instance key.
+        If the key is not in the cache, raises a KeyError.
+        If there's no Mor with the given name, raises a MorNotFoundError.
+        """
+        with self._mor_lock:
+            mors = self._mor[key]
+            try:
+                return mors[name]
+            except KeyError:
+                raise MorNotFoundError("Mor object '{}' is not in the cache.".format(name))
+
+    def set_metrics(self, key, name, metrics):
+        """
+        Store a list of metric identifiers for the given instance key and Mor
+        object name.
+        If the key is not in the cache, raises a KeyError.
+        If the Mor object is not in the cache, raises a MorNotFoundError
+        """
+        with self._mor_lock:
+            mor = self._mor[key].get(name)
+            if mor is None:
+                raise MorNotFoundError("Mor object '{}' is not in the cache.".format(name))
+            mor['metrics'] = metrics
+
+    def mors(self, key):
+        """
+        Generator returning all the mors in the cache for the given instance key.
+        """
+        with self._mor_lock:
+            for k, v in self._mor.get(key, {}).iteritems():
+                yield k, v
+
+    def mors_batch(self, key, batch_size):
+        """
+        """
+        with self._mor_lock:
+            mors_dict = self._mor.get(key)
```

Comment:
same question regarding `KeyError`

Reply:
Same consideration as above

Reply:
Also this generator is probably a bit counterintuitive, going to add an example in the docstring